### ISSUE TITLE
Potential energy inconsistency when creating new contexts with CUDA code

### ISSUE THREAD

**Issue by JIMonroe** [pandegroup/openmm#2191]
Hopefully this is just a fringe case, but I'm posting this to make sure.

When I run OpenMM (7.3.0 for python 2.7) using the CUDA platform (CUDA 8.0 with drivers 367.44) on Pascal GeForce GTX 1080 GPUs, I run into inconsistent potential energies when creating new contexts.  This seems to happen randomly and only when I'm working with alchemical systems (so I have CustomNonbondedForce objects).  It happens more frequently when more of the GPU memory is being used and has never happened with CPU or OpenCL platforms.  In the example code and outputs I've attached, I use test systems from OpenMMTools, but I have also reproduced this with other systems as well.  

As a little background, I ran into this while running alchemical simulations to compute free energies of solvation.  In order to save time at the expense of memory, I was creating a different context for every alchemical state to avoid re-computing long-range corrections.  Occasionally, one of the contexts would have a potential energy that was off right from the start and would stay off for the whole simulation.  This was despite the fact that the system object reference by the context would yield correct potential energies when copied and applied to a new context (i.e. I wrote an energy decomposition function based on those available in OpenMMTools alchemical factory and ParmEd).  I think the code provided is replicating this same issue with a minimal amount of code and some standard test systems.  If more test systems are needed, I can provide additional code and input files.

For now I have a work-around.  I have not tried updating my driver version (I know it's old), but may eventually get to that.

[CUDA_Inconsistency.tar.gz](https://github.com/pandegroup/openmm/files/2508640/CUDA_Inconsistency.tar.gz)

**Comment by jchodera** [comment]
The CUDA platform, when used with PME, produces non-deterministic energies and forces for purposes of performance. This can be disabled with a small performance hit by setting the `DeterministicForces` platform flag to `True`. Perhaps this is the cause of the effect you're seeing, since OpenCL and CPU platforms are deterministic?

@peastman: This seems to be hard to find in the OpenMM docs. Could we maybe make this more prominent in one or two places?

**Comment by JIMonroe** [comment]
Thank you!  That's really good to be aware of, but I'm not sure it's the issue.  I tried setting `DeterministicForces` to `True` and now I randomly get `NaN` instead of a number (which is honestly an improvement because it more clearly tells you something is going wrong).

I'm not sure in what way the CUDA platform is non-deterministic, but what I've seen is a ~100 kB*T offset in one of the contexts.  I was only able to focus in on the issue because I tried solvating united-atom methane and left in the alchemical states to decouple electrostatics, which resulted in some states with identical potential energies.  The potential energies I got from the contexts representing the different alchemical states were something like:

```
Time (ps)       Potential energies of alchemical states (kB*T)
0.0              -30000.00  -30100.00  -30000.00  -30000.00 ...
1.0              -29500.00  -29610.00  -29500.00  -29500.00 ...
```

In other words, once it randomly gets off, it stays off in the same way for the rest of the energy evaluations with different configurations.  I can send files to reproduce this as well, but it's quite a bit more code to look through and I think it's the same issue.

**Comment by andrrizzi** [comment]
This may be connected with the issue I've been debugging in YANK for a while that I described in #2171. In my case, the problem vanishes if I run with an OpenMM commit that is currently uploaded on anaconda in the `dev` channel as 7.5.0. It may be worth the shot trying it as well to see if it's the same commit that is causing the problem.

**Comment by peastman** [comment]
Are you using the 7.3 beta or the latest code from the repository?  There've been a couple of bugs fixed since the beta that might have caused this.  Could you try the latest dev build, or the latest code from the repository, and see if the problem still occurs?

**Comment by andrrizzi** [comment]
The latest dev build is currently posted as `7.4.0` in the `dev` channel (sorry for the confusion there, we've been temporarily uploading several builds to track down the problem).

**Comment by andrrizzi** [comment]
I gave it a go too, and I can reproduce the problem with the dev version (after 294 automatic trials of the script).

**Comment by peastman** [comment]
It ran through all 500 trials with no inconsistencies.  I've increased `nCopies` to 5 and I'm running it again.  It's up to about 60 trials with no problems yet.

I'm using CUDA 9.1 with the 387.26 driver on a Titan X.  I don't know if any of that makes a difference.  I compiled from source using the latest code in the repository.

**Comment by andrrizzi** [comment]
Not sure if it helps, but the script doesn't stop when it detects an inconsistency, so you need to read/search through the whole output. It got me at first.

**Comment by peastman** [comment]
I know.  I sat and watched it the whole time. :)

**Comment by JIMonroe** [comment]
Yeah, sorry about that... I wanted to know if it was just off for one iteration or stayed off.  When it's off it prints a message you can grep (search for "INCONSISTENCY").

After 1000 trials with the most recent 7.4.0 dev release I see no errors.  This is regardless of whether I use `DeterministicForces` set to `True` or `False`.  Whatever is different in the most recent dev version seems to have fixed this, unless it's just even more rare now.  I'm still running more trials.

**Comment by JIMonroe** [comment]
Ok, correction: in the dev version I still get the error if `DeterministicForces` is `False`.  I don't get the error if it's set to `True` (at least not in 1000 trials).

**Comment by peastman** [comment]
Or possibly it's nothing to do with that, and this just is a very intermittent issue...

**Comment by peastman** [comment]
Could you try a few experiments?  First, what if you change the NonbondedForce from PME to CutoffPeriodic.  Do you still get the error?

I also noticed this line in the script:

```python
systemCopy = copy.deepcopy(systemRef)
```

Is that actually needed to reproduce it?  What if you just use `systemRef` directly when creating the new contexts?

**Comment by peastman** [comment]
Still no errors after setting `DeterministicForces` and rerunning.

**Comment by JIMonroe** [comment]
I still get an inconsistency when I just reference `systemRef` instead of using copy.deepcopy.  With `CutoffPeriodic` I saw no issues.

**Comment by peastman** [comment]
I merged the fix.  Give it a try and see if it makes a difference.

**Comment by JIMonroe** [comment]
Thanks!  I'll give it a try, but this may take me a little time as I've never actually compiled OpenMM from source before.

**Comment by andrrizzi** [comment]
If you can wait a few hours, I believe an updated 7.4 should be automatically uploaded to anaconda.

**Comment by JIMonroe** [comment]
Thanks!  That made things easier.

Unfortunately, even with the new 7.4.0 dev build I'm still seeing the energy go off sometimes.  It's happened now with `DeterministicForces` set to both `True` and `False` (though as before, with `False` I get a number and with `True` I get NaN).

@andrrizzi are you still able to reproduce this?  And with what CUDA version, GPUs, and drivers?

**Comment by andrrizzi** [comment]
Confirming that I can still reproduce this with the latest 7.4, both with the script and with YANK, although it looks to me it takes longer to reproduce (it took 700-800 attempts). I've also observed the same `DeterministicForces` behavior. I always get a constant offset when `DeterministicForces` is `False` and NaN when it's `True`.

**Comment by peastman** [comment]
Did you try switching from PME to CutoffPeriodic?

**Comment by andrrizzi** [comment]
Trying now.

**Comment by andrrizzi** [comment]
I couldn't reproduce it after ~2700 trials. I'm going to check if the issue disappears with commit dbeb0c2 . If this is the case, this is very likely the same issue we've found with YANK, which we identified being introduced either with 1b12ede or 36b9db8 .

**Comment by peastman** [comment]
> I couldn't reproduce it after ~2700 trials.

Is that referring to when you switched to CutoffPeriodic?

**Comment by andrrizzi** [comment]
> Is that referring to when you switched to CutoffPeriodic?

Yes, sorry (modified the comment now).

**Comment by peastman** [comment]
Thanks, those are good clues.  Are you certain it's actually using CPU PME?  Check `context.getPlatform().getPropertyValue(context, "UseCpuPme")`.  If for some reason the CPU PME plugin can't be loaded it will just ignore your request to use it and run PME on the GPU as usual.

**Comment by JIMonroe** [comment]
Yeah, good call.  I'm getting `False`, so it's not actually doing the PME calc on the CPU.

Can someone else check this, or is there a way I can get this feature working?  The CPUs are SSE 4.1 compatible, but I'm not certain my FFTW install is as OpenMM would like it.

**Comment by peastman** [comment]
I guessed that was probably the case, since all other signs point to this being in the CUDA PME code.  Try

`ldd (OpenMM plugins directory)/libOpenMMPME.so`

That will tell you if there's a dependency it can't resolve.  Also take a look at `Platform.getPluginLoadFailures()`.

**Comment by andrrizzi** [comment]
Here's the results:

- I can't reproduce the problem with 7.2.2.
- Up to commit dbeb0c2 (before the CUDA PME optimization), I can reproduce the problem, but I always get a constant offset with either `DeterministicForces` set to `True` or `False.
- After commit 36b9db8 (the CUDA PME optimization), I can reproduce the problem, but in this case setting `DeterministicForces` to `True` results in a NaN instead of a constant offset. This is why we caught it in YANK.

I can try to backtrack the faulty commit now that we have a simple script that reproduces the problem as the CUDA PME optimization doesn't look to be at the bottom of this.

**Comment by andrrizzi** [comment]
I'm testing commits a8020adc16669eb4d4d2ad9fe77d453ed3664469, b441c4a7b74fc260d1dee125e9f7ca69ff0bdfb4, and c0e862f61ba60a877627f4880a1545918add4ad2 next, where some changes to the CUDA code were introduced.

**Comment by andrrizzi** [comment]
Also, I forgot to add that the constant offset is always the same between executions. The energy of the configuration always goes from -6301 to -6844.
```
On trial 472...
True energy before copy: -6301.885414
Copy energies: [-6844.27722018]
```

**Comment by JIMonroe** [comment]
So this might narrow down the bug...  I got `UseCpuPme` working, but it only works with OpenCL.  Noting this, I just ran through and tried all the platform options I saw related to PME with CUDA and the results are as follows:

- With `UseCpuPme` set to `True`, I get random Seg faults, which happen when I call `context.getState(getEnergy=True).getPotentialEnergy()`.
- The Seg fault with `UseCpuPme` set to `True` happens for any system, regardless of whether or not it is alchemical, with `CustomNonbondedForce` objects.
- If `CudaDisablePmeStream` is set to `True`, the Seg faults go away and everything runs perfectly with no errors (no Seg fault, no offset, no NaN) in the 7.4.0 dev version.

I'm currently checking to see if this is completely independent of the other platform options and to see if it fixes problems with the 7.3.0 version.

**Comment by andrrizzi** [comment]
Hmm. I compiled the latest OpenMM `master` instead of using a package pushed on anaconda, and I can't reproduce the problem anymore. Is it possible this has something to do with the compilation of the conda builds? @peastman you mentioned you didn't observe any failure. Were you using a locally compiled installation or a conda package?

**Comment by peastman** [comment]
All my testing has been compiling from source.

**Comment by jchodera** [comment]
@peastman : It sounds like it would be useful for you to try the conda package built from `master` to compare:
```bash
conda clean -tipsy
conda install --yes -c omnia/label/devcudaXY openmm==7.4.0
```

**Comment by peastman** [comment]
> If CudaDisablePmeStream is set to True, the Seg faults go away and everything runs perfectly 

Is that with or without UseCpuPme?  That is, if you do *not* use CPU PME, and you *do* disable the PME stream, are all results correct?

(If so, I'm confused.  The revisions that John bracketed shouldn't have anything to do with the PME stream.  On the other hand, if you're doing PME on the CPU the `DisablePmeStream` option should have no effect on anything.)

**Comment by JIMonroe** [comment]
> Is that with or without UseCpuPme? That is, if you do not use CPU PME, and you do disable the PME stream, are all results correct?

Yes.  And that's true for all versions I've tried now (including 7.3.0).  The way I set up the platform is as follows:

```
platform` = mm.Platform.getPlatformByName('CUDA')
prop = {'Precision': 'mixed', 
             'DeviceIndex': '0',
             'DeterministicForces': 'False',
             'UseCpuPme': 'True',
             'CudaDisablePmeStream': 'True', 
            }
```

If I run as above, or with `UseCpuPme` set to `False`, I see no errors (not after ~1000 iterations with the number of copies set to 5).

If I switch `CudaDisablePmeStream` to `False` while `UseCpuPme` is True, I get Seg faults.  

If `CudaDisablePmeStream` is `False` while `UseCpuPme` is also `False`, I get the original error. 

(I've tried all this with `DeterministicForces` set both ways).

I'm also running `context.getPlatform().getPropertyValue(context, "UseCpuPme")` and `context.getPlatform().getPropertyValue(context, "CudaDisablePmeStream")` to make sure they are both getting set and they are.  I'm sorry I can't offer much insight as to what's going on, but these are the results I'm finding (and I'm not ruling out a funky install on my part - that's certainly possible).

**Comment by andrrizzi** [comment]
Confirming that I can reproduce the original problem with the conda 7.4, but I can't with the local installation. However, I can reproduce an instantaneous segfault with the local installation using these properties
```
prop = {
    'Precision': 'mixed', 
    'DeviceIndex': '0', 
    'DeterministicForces': 'false',
    'UseCpuPme': 'true',
    'CudaDisablePmeStream': 'false'
}
```

**Comment by peastman** [comment]
I think the segfault is probably unrelated.  I'm not going to worry about that right now, since CPU PME is a very rarely used feature.  Knowing the error goes away when you disable the PME stream seems like a good clue.  I'll try to follow that.

**Comment by peastman** [comment]
I'm trying to install the dev build into a new environment, but it won't let me install `openmmtools`.

```
$ conda install -c conda-forge -c omnia openmmtools
Solving environment: failed

UnsatisfiableError: The following specifications were found to be in conflict:
  - openmmtools -> mdtraj -> python=2.7
  - python=3.7
Use "conda info <package>" to see the dependencies for each package.
```

I'm using Python 3.6.

**Comment by andrrizzi** [comment]
Hmm. I just tried `conda create --name testenv openmmtools python=3.6` and it was able to solve the package specifications for me. I'm not sure why it's complaining about py 2.7.

**Comment by peastman** [comment]
I've come up with a hypothesis for what might be going on.  It doesn't actually fit all the facts, but it fits many of them, so I figured I'd at least try a potential fix.  It's in #2193.  Could you give it a try?

**Comment by andrrizzi** [comment]
I tried it, but unfortunately, I can still reproduce. I'll go back to try previous commits.

**Comment by peastman** [comment]
Sigh.  I wish I could reproduce it.  Without that, I'm just stumbling in the dark.  And it's really strange that you also can't reproduce it when you compile from source.  I don't know what to make of that.

Ok, let's try something else.  All the discussion so far has talked about energy.  What about forces?  Do they also change or become nan, or is it just the energy?

**Comment by jchodera** [comment]
> Sigh. I wish I could reproduce it. Without that, I'm just stumbling in the dark. And it's really strange that you also can't reproduce it when you compile from source. I don't know what to make of that.

Are you still not able to install 7.4.0 from `omnia/label/devcudaXY`? 
You can start from a fresh miniconda installation:
```bash
# install miniconda for linux
wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh -b -p miniconda
export PATH=./miniconda:$PATH
# install openmm and openmmtools
conda install -c omnia/label/devcuda100 -c omnia -c conda-forge openmm==7.4.0 openmmtools
```

**Comment by peastman** [comment]
> Are you still not able to install 7.4.0 from omnia/label/devcudaXY?

No.  See the error message above.

Even if I could reproduce it with the conda package, it would be very difficult to do any debugging with it because I couldn't modify it.

**Comment by jchodera** [comment]
> No. See the error message above.

Ah! I know how to fix that, but it will be after I migrate the `conda-recipes` build framework.

@andrrizzi was correct that you need to force it to not use Python 3.7 right now, but 3.6 or 3.5:
```
conda create --name testenv -c omnia/label/devcuda100 -c omnia -c conda-forge openmmtools python=3.6 openmm==7.4.0
```

**Comment by jchodera** [comment]
> Even if I could reproduce it with the conda package, it would be very difficult to do any debugging with it because I couldn't modify it.

I think just knowing whether there really *is* a difference in behavior between the exact same git commit built locally and via our docker image would be very valuable! @andrrizzi is suggesting that this might be the case, I think.

**Comment by peastman** [comment]
> I think just knowing whether there really is a difference in behavior between the exact same git commit built locally and via our docker image would be very valuable!

I already know that's the case, because @andrrizzi observed it.  But I can't infer much from that observation.  If the problem is uninitialized memory, any change that causes even a tiny shift in memory layout could cause it to appear or disappear.  For example, using a different compiler.  Likewise if it's a race condition, any change that causes a slight difference in relative timings could cause it to appear or disappear.

That's why, even though you identified a narrow window in the commit history where the problem started happening, I can't be certain it's actually caused by an error introduced in those commits.  They might have just perturbed internal state or timings in a way that caused a previously existing error to start causing problems.

The only reliable way to debug something like this is to catch it in the process of happening, then trace back to find the moment where something first goes wrong.

**Comment by andrrizzi** [comment]
I haven't been able to reproduce the problem with a8020ad . I'm now trying the next two commits b441c4a7b74fc260d1dee125e9f7ca69ff0bdfb4 and c0e862f61ba60a877627f4880a1545918add4ad2 . The latter adds the first implementation of the offset parameters. Let's see if we can get more clues to where to look at at least.

It is a very long process, but I think it may be possible to "modify" the conda version by using a test branch to push changes.

**Comment by JIMonroe** [comment]
@peastman I'm sorry I can't help you reproduce this, but in the mean time I'll try to keep providing information that may help narrow things down when you are able to.

> Ok, let's try something else. All the discussion so far has talked about energy. What about forces? Do they also change or become nan, or is it just the energy?

I checked the forces in addition to the energies:

- With `DeterministicForces` set to `False` the offset in potential energies is indeed accompanied by an offset in the forces.
- Interestingly, with `DeterministicForces` set to `True`, the energies go to NaN, but the forces shift by the same offset as with this key set to `False`.  So the energies are NaN, but the forces are not.
- I know you're ignoring it, but with `UseCpuPme` set to `True`, I get a Seg fault as soon as I call context.getState(), but ONLY if I ask it for forces or energies (positions, velocities, and parameters are fine).

**Comment by peastman** [comment]
> With DeterministicForces set to False the offset in potential energies is indeed accompanied by an offset in the forces.

Thanks, that's good to know.  It means the error is somewhere that affects both forces and energies, most likely in the charge grid.

**Comment by andrrizzi** [comment]
Here's a summary of my experiments with different commits (from the older to the newest).
- a8020ad and b441c4a : could _not_ reproduce the problem.
- c0e862f : could reproduce the constant offset. This commit is right after b441c4a and it's a "merge from master" commit introducing in the code the parameter offset feature.
- dbeb0c2 still always fails with a constant offset in the energies, and two commits later 36b9db8 fails with NaN when `DeterministicForces` is set to `true`.

At this point, it becomes a little trickier to narrow in because the commits implementing the offset feature are somewhat partial, but I'm going to test 7ebaa81 and 7c6f0b4 (i.e. right before and after implementing the offset feature on the CUDA platform).

**Comment by andrrizzi** [comment]
Sorry for the confusion. I said that when I was looking only for NaNs instead of a constant offset in the energy. The NaN indeed appears only after 1b12ede or 36b9db8. I was always using `DeterministicForces = true` at that point so I didn't catch the other error.

> You also said you couldn't reproduce the problem when building from source. So how are you testing these revisions?

We're temporarily pushing specific commits on the dev channel on anaconda.

**Comment by peastman** [comment]
Perhaps those commits were a red herring and I've been looking at the wrong code.  Is it possible for you to create dev builds from code you modify yourself, not just points in the actual OpenMM revision history?  If so, could you try an experiment?  Line 2097 of CudaKernels.cpp is

```c++
if (recomputeParams || hasOffsets) {
```

Just before that line could you insert

```c++
recomputeParams = true;
```

so the block will always be executed?  I predict one of three things will happen: the error will go away, the error will be easier to reproduce, or it will have no effect. :)

**Comment by andrrizzi** [comment]
Unfortunately, I think it had no effect. I will create a temporary PR where you can push changes as well. I'm uploading the package on anaconda under the version `openmm=debug` in the `dev` channel.

**Comment by peastman** [comment]
#2195 is yet another attempt at a change that might or might not fix this.

**Comment by peastman** [comment]
To put everything in one place, let me summarize what we think we know about this problem (keeping in mind that anything we think we know could turn out to be wrong!).

- The problem only occurs if you use PME.
- Furthermore, it only occurs if you run PME in a separate stream.
- The behavior is different if you enable deterministic forces.  The only effect of that flag is on how PME charge spreading gets done.

All of that indicates pretty strongly that this is somehow related to PME.  But continuing:

- The problem first appeared in commit c0e862f61ba60a877627f4880a1545918add4ad2.  That was the one that introduced offset parameters.

The test case doesn't use offset parameters, and very little of that commit had anything to do with PME.  As far as I can tell, the only connection is that it changed where and how charges get recorded.  It used to be done in host code that directly copied them to their final locations.  That commit changed to having a GPU kernel do the copying.

This still has little to do with the PME stream.  The one connection I can see is that it has to perform a stream synchronization after executing the kernel:

```c++
if (usePmeStream) {
    cuEventRecord(pmeSyncEvent, cu.getCurrentStream());
    cuStreamWaitEvent(pmeStream, pmeSyncEvent, 0);
}
```

One other possibly significant fact:

- The error has only been observed with CUDA, not OpenCL.

Since the two platforms are very similar, that suggests the problem might (or might not) be in one of the few pieces of code that varies between them.

Anyway, those are the observations that led me to #2195.  We'll see whether that's really it or not.

**Comment by andrrizzi** [comment]
After the #1295 fix, I have been able to reproduce the problem neither with the script nor with YANK since yesterday evening.

@JIMonroe, could you try too? You can install the fix with `conda install openmm=debug -c omnia/label/dev --force`.

**Comment by peastman** [comment]
Yay!!!!!

I'll merge the fix.  That was quite an ordeal to track down!

**Comment by andrrizzi** [comment]
Uff, one out of 4 YANK simulations crashed with the same problem after running for 4 hours T.T . It may be just much harder to reproduce.

**Comment by peastman** [comment]
Oh well, that means there's still *something* wrong *somewhere*.  But I think we should still go ahead and build the release candidate.  If you only saw one crash in 16 hours of running simulations, that's going to be very hard to track down, and I don't want to keep holding up the many important fixes already in 7.3.

**Comment by JIMonroe** [comment]
Hmm, I installed the debug version of OpenMM from conda and actually still see the error in the original script.  _Maybe_ I had to run for a little longer to see it, but I'm still getting all the same behaviors as before.

**Comment by peastman** [comment]
What does `version.git_revision` equal?

**Comment by JIMonroe** [comment]
When I look at `conda list`, the version is shown as "debug".

openmm.version.git_revision shows f03d9ae2fb31ae2d60c1f41b45fd401a6cbe4d10
openmm.version.full_version shows 7.3.0.dev-f03d9ae

These are both in python after importing openmm from simtk.

**Comment by peastman** [comment]
That's the correct one.  @andrrizzi are you still not able to reproduce it with @JIMonroe's script?

**Comment by andrrizzi** [comment]
Nope! I haven't been able to reproduce it with the script. Only with YANK.

**Comment by peastman** [comment]
Now I'm really confused!  Why would that change fix it for one person but not another?  I really don't know what's going on here.

**Comment by peastman** [comment]
@JIMonroe is there any way it could still be using the old version when running the script?  That could happen, for example, if you have two copies of OpenMM on your computer and `OPENMM_PLUGIN_DIR` is pointing to the other one.  Also check both `version.openmm_library_path` and `Platform.getDefaultPluginsDirectory()`.  I'm just really confused about what's going on, and it still seems to me like the most plausible explanation is that it's somehow using the old version of the plugin without the fix on your computer.

**Comment by JIMonroe** [comment]
The machine I'm on isn't connected to the internet, so my procedure to change out OpenMM has been `conda uninstall openmm`, then running `conda install openmm_some_version.tar.bz2` after I've downloaded the appropriate tarball from Anaconda Cloud.  I've checked dependencies and _think_ I'm fine, but there are a couple (maybe) weird things below:

`OPENMM_PLUGIN_DIR` is not assigned - not sure if this is an issue.
`openmm.version.openmm_library_path` returns `/home/jmonroe/bin/anaconda2/lib`, which seems right.
`openmm.Platform.getDefaultPluginsDirectory()` returns `/usr/local/openmm/lib/plugins`, which is weird because that directory does not exist, and besides, I've done a local install and should not have write access to that directory without switching to an admin account with super-user privileges.

Sorry for the delay on answering this!  I'm running more tests with the 7.3.0 release candidate version, and hopefully that works out.

**Comment by jchodera** [comment]
@peastman: We're still struggling with this hard-to-track-down NaN bug.

Do you think any of the `*Sanitizer` compilation flags for `clang` could help track this down?
https://clang.llvm.org/docs/index.html

There are a bunch of things to try that slow down performance by a bit but should identify (CPU) memory, thread, and address issues:
```
Thread Safety Analysis
AddressSanitizer
ThreadSanitizer
MemorySanitizer
UndefinedBehaviorSanitizer
DataFlowSanitizer
LeakSanitizer
```
I've been trying to adjust the `dev` builds in [this PR](https://github.com/omnia-md/conda-dev-recipes/pull/157) but have run into trouble since these must be compiled in dynamically but our `conda` recipe seems to want to compile them in statically (?!).

**Comment by JIMonroe** [comment]
Well, not sure why, but I do still see potential energy inconsistencies with the 7.3.0 release candidate and the original script.  The errors seem to only happen ~1/1000 times now, so less frequently, so it's taken me longer to run.  Changes to all the different platform settings still affect things in the same way.  Again, I am not sure if this is specific to the combination of drivers/CUDA version/GPUs that I am using, but am happy to test out any suggestions anyone has.

One observation I would like to make is that in the script I provided, once the potential energy is off, ALL of the potential energies remain off.  To me this means the original system object is somehow being corrupted, because this is what is being repeatedly referenced when constructing new contexts.

**Comment by peastman** [comment]
I tried running on a different computer to see if that would make a difference.  I tried both the release candidate and a version built from source.  I still haven't seen a single error. :(

**Comment by peastman** [comment]
@JIMonroe and @andrrizzi this is a bit of a long shot, but could you try changing the nonbonded method from PME to LJPME?  I believe you can do that by just adding the option `nonbondedMethod=app.LJPME` to the `AlchemicalWaterBox` constructor.  Does that make the error go away?

**Comment by peastman** [comment]
I've merged another fix (#2205).  Once again I don't know if it will help, but give it a try.  Or just try LJPME as I suggested above, since the error didn't happen in that case.

**Comment by andrrizzi** [comment]
After #2205, I couldn't reproduce the NaNs with YANK after 12 x 6h of calculations! I'm going to try @JIMonroe's script too today, although I already had trouble reproducing the problem with it.

@JIMonroe, if you want to give it a go, note that the development version (7.4) is now pushed on the [omnia-dev](https://anaconda.org/omnia-dev/openmm/files) channel, not omnia anymore.

**Comment by peastman** [comment]
Yay!

**Comment by andrrizzi** [comment]
I couldn't reproduce with @JIMonroe's script after 4000 attempts.

**Comment by peastman** [comment]
Yay again!  Then with great relief I am finally going to close this issue!

**Comment by jchodera** [comment]
@peastman : I'd love to explore how we could augment the travis tests with an profiling or leak-detection tests that would allow us to catch these problems earlier. This particular issue set my group back by a month, so I'm highly motivated to see how we could spend a bit of time to catch problems in this class early and prevent them from cropping up again!

**Comment by peastman** [comment]
If you have any ideas, I'd be glad to hear them.  cuda-memcheck and valgrind are both helpful tools, but very slow to run and with a high rate of false positives.  Beyond that I don't have a lot of suggestions.

**Comment by jchodera** [comment]
The clang compiler provides a bunch of sanitizers that only slow things down by about 2x, such as:

https://clang.llvm.org/docs/MemorySanitizer.html

Do you have any experience with these?

**Comment by peastman** [comment]
No, I've never used them.
